From 051aad96d87cbbce464990584663cc5a0a867ec1 Mon Sep 17 00:00:00 2001
From: Mathieu Tortuyaux <mtortuyaux@microsoft.com>
Date: Thu, 20 Aug 2026 14:23:22 +0200
Subject: [PATCH 03/15] config/v3_7: convert ignition 2.x to 3.x

if the version of ignition is 2.x we convert it to 3.x using
ign-converter.

it should support any 2.x version

Signed-off-by: Mathieu Tortuyaux <mtortuyaux@microsoft.com>
---
 config/v3_7_experimental/config.go | 37 ++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/config/v3_7_experimental/config.go b/config/v3_7_experimental/config.go
index c82ca703..57c958c7 100644
--- a/config/v3_7_experimental/config.go
+++ b/config/v3_7_experimental/config.go
@@ -15,9 +15,14 @@
 package v3_7_experimental
 
 import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/coreos/ignition/config/v2_4"
 	"github.com/coreos/ignition/v2/config/merge"
 	"github.com/coreos/ignition/v2/config/shared/errors"
 	"github.com/coreos/ignition/v2/config/util"
+	"github.com/coreos/ignition/v2/config/v24tov31"
 	prev "github.com/coreos/ignition/v2/config/v3_6"
 	"github.com/coreos/ignition/v2/config/v3_7_experimental/translate"
 	"github.com/coreos/ignition/v2/config/v3_7_experimental/types"
@@ -67,6 +72,38 @@ func ParseCompatibleVersion(raw []byte) (types.Config, report.Report, error) {
 		return types.Config{}, rpt, err
 	}
 
+	// if the version is 2.x or 1.x, we
+	// convert it to 3.1
+	if version.Major != 3 {
+		// Parse should fallback on every 2.x supported version
+		cfg, _, err := v2_4.Parse(raw)
+		if err != nil || rpt.IsFatal() {
+			return types.Config{}, report.Report{}, fmt.Errorf("unable to parse 2.x ignition: %w", err)
+		}
+
+		/*
+			map[string]string{} is required by the ign-converter
+			Ignition Spec 3 will mount filesystems at the mountpoint specified by path when running.
+			Filesystems no longer have the name field and files, links, and directories no longer specify the filesystem by name.
+			This means to translate filesystems (with the exception of root),
+			you must also provide a mapping of filesystem name to absolute path, e.g.
+			```
+			map[string]string{"var": "/var"}
+			```
+		*/
+		newCfg, err := v24tov31.Translate(cfg, map[string]string{})
+		if err != nil {
+			return types.Config{}, report.Report{}, fmt.Errorf("unable to translate 2.x ignition to 3.1: %w", err)
+
+		}
+
+		// update raw in place to continue with the 3.x logic
+		raw, err = json.Marshal(newCfg)
+		if err != nil {
+			return types.Config{}, report.Report{}, fmt.Errorf("unable to render JSON: %w", err)
+		}
+	}
+
 	if version == types.MaxVersion {
 		return Parse(raw)
 	}
-- 
2.55.0

